ledger-core/virtual/execute/execute_test.go

```diff
@@ -563,4 +564,99 @@ func TestSendVStateReportWithMissingState_IfConstructorWasInterruptedBeforeRunne
 		t.Error("timeout")
 	}
+
+	mc.Finish()
+}
+
+func TestSMExecute_StopWithoutMessagesIfPulseChangedBeforeOutgoing(t *testing.T) {
+	t.Log("C5101")
+	defer testutils.LeakTester(t,
+		goleak.IgnoreTopFunction("github.com/insolar/assured-ledger/ledger-core/runner.(*worker).Run.func1"),
+		goleak.IgnoreTopFunction("github.com/insolar/assured-ledger/ledger-core/conveyor/smachine.startChannelWorkerUnlimParallel.func1"),
+	)
+
+	var (
+		mc  = minimock.NewController(t)
+		ctx = context.Background()
+
+		class     = gen.UniqueGlobalRef()
+		caller    = gen.UniqueGlobalRef()
+		objectRef = gen.UniqueGlobalRef()
+
+		catalog     object.Catalog         = object.NewLocalCatalog()
+		authService authentication.Service = authentication.NewServiceMock(t)
+	)
+
+	slotMachine := slotdebugger.New(ctx, t)
+	slotMachine.PrepareMockedMessageSender(mc)
+	slotMachine.PrepareMockedRunner(ctx, mc)
+	slotMachine.AddInterfaceDependency(&catalog)
+	slotMachine.AddInterfaceDependency(&authService)
+
+	var vStateReportRecv = make(chan struct{})
+	checkMessage := func(msg payload.Marshaler) {
+		res, ok := msg.(*payload.VStateReport)
+		require.True(t, ok)
+		assert.Equal(t, objectRef, res.Object)
+		assert.Equal(t, payload.Ready, res.Status)
+		assert.Equal(t, int32(0), res.OrderedPendingCount)
+		assert.Equal(t, int32(0), res.UnorderedPendingCount)
+		assert.Equal(t, []byte("213"), res.ProvidedContent.LatestDirtyState.State)
+		close(vStateReportRecv)
+	}
+	slotMachine.MessageSender.SendRole.SetCheckMessage(checkMessage)
+
+	outgoing := slotMachine.GenerateLocal()
+
+	slotMachine.RunnerMock.AddExecutionClassify(
+		reference.NewRecordOf(objectRef, outgoing).String(),
+		contract.MethodIsolation{Interference: contract.CallTolerable, State: contract.CallDirty},
+		nil,
+	)
+
+	smObject := object.NewStateMachineObject(objectRef)
+	smObject.SetState(object.HasState)
+	smObject.SetDescriptor(descriptor.NewObject(reference.Global{}, reference.Local{}, class, []byte("213"), reference.Global{}))
+	slotMachine.AddStateMachine(ctx, smObject)
+
+	smExecute := SMExecute{
+		Payload: &payload.VCallRequest{
+			CallType:            payload.CTMethod,
+			Caller:              caller,
+			Callee:              objectRef,
+			CallFlags:           payload.BuildCallFlags(contract.CallTolerable, contract.CallDirty),
+			CallOutgoing:        outgoing,
+			CallSiteDeclaration: class,
+			CallSiteMethod:      "test",
+			Arguments:           insolar.MustSerialize([]interface{}{}),
+		},
+		Meta: &payload.Meta{
+			Sender: caller,
+		},
+	}
+	slotMachine.Start()
+	defer slotMachine.Stop()
+
+	smWrapper := slotMachine.AddStateMachine(ctx, &smExecute)
+	slotMachine.RunTil(smWrapper.BeforeStep(smExecute.stepStartRequestProcessing))
+	slotMachine.Migrate()
+	slotMachine.RunTil(smWrapper.AfterStop())
+
+	go slotMachine.RunTil(func(event debuglogger.UpdateEvent) bool {
+		select {
+		case <-vStateReportRecv:
+			return true
+		default:
+			return false
+		}
+	})
+
+	select {
+	case <-vStateReportRecv:
```

Comment:
we have testutils.WaitSignalsTimed for that

Reply:
done